### DISCUSSION TITLE
Swell Client - TypeScript - Incomplete Types

### DISCUSSION THREAD

**Discussion by Lulololu** [swellstores/community#325] [Help] [answered]
**Hi,**

We switched from the Graphql client to the Swell rest library, [because not all features were available with GraphQL](https://github.com/orgs/swellstores/discussions/323#discussioncomment-6553369). But we're having some difficulties with Typescript.

So we were wondering if, for some reason, the swell client methods, [as described in the Frontend documentation](https://developers.swell.is/frontend-api/products), should not be used in some case.

**1 - For example, whereas we get a pageCount as query parameter with the GraphQL client, TypeScript does not allow a **page_count** result with the Rest client, even if we do receive it as result.**

```
    const initialData = await swellClient.categories.list({ $locale: locale, limit: 100 });
    const pageCount = initialData.page_count;
```

```
Property  page_count  does not exist on type
OperationResult<CategoriesQuery, Exact<{ search: string; where?: any; sort?: InputMaybe<string> | undefined; limit?: InputMaybe<number> | undefined; page?: InputMaybe<...> | undefined; _locale?: InputMaybe<...> | undefined; _currency?: InputMaybe<...> | undefined; }>>
```

<img width="155" alt="image" src="https://github.com/swellstores/community/assets/73346696/3e8c9c95-9375-415a-97b3-7972d145bf05">

**2 - Same thing happens with products.list, where we can't use the locale if we listen to TypeScript:**

`swellClient.products.list({
        $locale: locale
    })`

`Object literal may only specify known properties, and  $locale  does not exist in type  ProductQuery `

Still, we get the localized results...

**3 - On the contrary, passing the locale arg [as described in the Backend documentation](https://developers.swell.is/backend-api/localization/retrieve-locale-values#retrieve-locale-values) doesn't rise any TS error**... because it's not typed. So we lose the benefit of Typescript. Also, what would be the point of using method described in the backend documentation when we are on the frontend ? Moreover, would it be considered as a bad practice (eventually, for security concerns)?

**Thanks for reading me,
Thanks for your work,
Thanks for your help,**

**Comment by logeshswell** [comment, marked as answer]
> TypeScript does not allow a `page_count` result with the Rest client.

@Lulololu No, you could fetch the count using `initialData.count`
**Code references:**
https://github.com/swellstores/swell-js/blob/master/types/index.d.ts#L173
https://github.com/swellstores/swell-js/blob/master/types/index.d.ts#L88

> We can't use the locale if we listen to TypeScript
> passing the locale arg as described in the Backend documentation doesn't rise any TS error

This is an incorrect implementation. You can't use `$locale` argument in swell-js while it would work fine in swell-node since it is a backend API specification and NOT because it's not typed.

Please refer to [Localization page](https://developers.swell.is/frontend-api/localization) for a proper implementation in swell-js with typescript.

> Would it be considered as a bad practice (eventually, for security concerns)?

Hmm, not really. But, you can use it if required. It is important to note that all the frontend APIs will pass through the backend APIs with limited functionalities. Hence, you receive the localized results using `$locale` field.
